fast64_internal/oot/exporter/room/shape.py

```diff
@@ -0,0 +1,350 @@
+
+
+@dataclass
+class RoomShapeImageSingle(RoomShapeImageBase):
+    """This class defines a room shape using only one image"""
+
+    imgName: str
+    width: int
+    height: int
+    otherModeFlags: str
+
+    unk_0C: int = 0
+    tlut: str = "NULL"
+    format: str = "G_IM_FMT_RGBA"
+    size: str = "G_IM_SIZ_16b"
+    tlutCount: int = 0
+
+    def getC(self):
+        """Returns the single background image mode variable"""
+
+        infoData = CData()
+        listName = f"RoomShapeImageSingle {self.name}"
+
+        # .h
+        infoData.header = f"extern {listName};\n"
+
+        # .c
+        infoData.source = (listName + " = {\n") + f",\n{indent}".join(
+            [
+                "{ " + f"{self.type}, {self.amountType}, &{self.entryArrayName}" + " }",
+                f"{self.imgName}",
+                f"0x{self.unk_0C:08X}",
+                f"{self.tlut}",
+                f"{self.width}, {self.height}",
+                f"{self.format}, {self.size}",
+                f"{self.otherModeFlags}, 0x{self.tlutCount:04X}",
+            ]
+        )
+
+        return infoData
+
+
+@dataclass
+class RoomShapeImageMulti(RoomShapeImageBase):
+    """This class defines a room shape using multiple images"""
+
+    bgEntryArrayName: str
+
+    def getC(self):
+        """Returns the multiple background image mode variable"""
+
+        infoData = CData()
+        listName = f"RoomShapeImageSingle {self.name}"
+
+        # .h
+        infoData.header = f"extern {listName};\n"
+
+        # .c
+        infoData.source = (listName + " = {\n") + f",\n{indent}".join(
+            [
+                "{ " + f"{self.type}, {self.amountType}, &{self.entryArrayName}" + " }",
+                f"ARRAY_COUNT({self.bgEntryArrayName})",
+                f"{self.bgEntryArrayName}",
+            ]
+        )
+
+        return infoData
+
+
+@dataclass
+class RoomShapeNormal:
+    """This class defines a normal room shape"""
+
+    name: str
+    type: str
+    entryArrayName: str
+
+    def getC(self):
+        """Returns the C data for the room shape"""
+
+        infoData = CData()
+        listName = f"RoomShapeNormal {self.name}"
+
+        # .h
+        infoData.header = f"extern {listName};\n"
+
+        # .c
+        numEntries = f"ARRAY_COUNT({self.entryArrayName})"
+        infoData.source = (
+            (listName + " = {\n" + indent)
+            + f",\n{indent}".join(
+                [f"{self.type}", numEntries, f"{self.entryArrayName}", f"{self.entryArrayName} + {numEntries}"]
+            )
+            + "\n};\n\n"
+        )
+
+        return infoData
+
+
+@dataclass
+class RoomShape(RoomShapeBase):
+    """This class hosts every type of room shape"""
+
+    sceneName: str
+    roomName: str
+
+    dl: Optional[RoomShapeDLists] = None
+    normal: Optional[RoomShapeNormal] = None
+    single: Optional[RoomShapeImageSingle] = None
+    multiImg: Optional[RoomShapeImageMultiBg] = None
+    multi: Optional[RoomShapeImageMulti] = None
```

Comment:
it's been a while but iirc it's just to keep room shape as a sub-system like everything else, this way if you wanted you could add a button that would only export this, this whole exporter was made based on this idea (which I implemented here for cutscenes and collisions)

Reply:
Sorry I'm not understanding, what about that prevents using different subclasses instead of having one class that switches its functionality based on type?

Reply:
I wrote that based on decomp structs, also what I said earlier too to be consistent

also I don't understand that question, what's the concern here exactly? 🤔